// @api.video/react-player@1.1.0/dist/index.js
'use strict';

var playerSdk = require('@api.video/player-sdk');
var React = require('react');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol */


var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var ApiVideoPlayer = React.forwardRef(function (props, ref) {
    var _a = React.useState(null), playerSdk$1 = _a[0], setPlayerSdk = _a[1];
    var iframeRef = React.useRef(null);
    var _b = React.useState(null), ratio = _b[0], setRatio = _b[1];
    var _c = React.useState([]), controls = _c[0], setControls = _c[1];
    var _d = React.useState(null), theme = _d[0], setTheme = _d[1];
    React.useEffect(function () {
        if (!iframeRef.current || !!playerSdk$1)
            return;
        var sdk = new playerSdk.PlayerSdk(iframeRef.current, __assign(__assign({}, props), props.video));
        sdk.addEventListener("play", function () { var _a; return (_a = props === null || props === void 0 ? void 0 : props.onPlay) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("pause", function () { var _a; return (_a = props.onPause) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("controlsdisabled", function () { var _a; return (_a = props.onControlsDisabled) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("controlsenabled", function () { var _a; return (_a = props.onControlsEnabled) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("ended", function () { var _a; return (_a = props.onEnded) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("error", function () { var _a; return (_a = props.onError) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("firstplay", function () { var _a; return (_a = props.onFirstPlay) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("fullscreenchange", function () { var _a; return (_a = props.onFullscreenChange) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("mouseenter", function () { var _a; return (_a = props.onMouseEnter) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("mouseleave", function () { var _a; return (_a = props.onMouseLeave) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("playerresize", function () { var _a; return (_a = props.onPlayerResize) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("qualitychange", function (v) { var _a; return (_a = props.onQualityChange) === null || _a === void 0 ? void 0 : _a.call(props, v.resolution); });
        sdk.addEventListener("ratechange", function () { var _a; return (_a = props.onRateChange) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("ready", function () { var _a; return (_a = props.onReady) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("resize", function () { var _a; return (_a = props.onResize) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("seeking", function () { var _a; return (_a = props.onSeeking) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("timeupdate", function (v) { var _a; return (_a = props.onTimeUpdate) === null || _a === void 0 ? void 0 : _a.call(props, v.currentTime); });
        sdk.addEventListener("useractive", function () { var _a; return (_a = props.onUserActive) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("userinactive", function () { var _a; return (_a = props.onUserInactive) === null || _a === void 0 ? void 0 : _a.call(props); });
        sdk.addEventListener("volumechange", function (v) { var _a; return (_a = props.onVolumeChange) === null || _a === void 0 ? void 0 : _a.call(props, v.volume); });
        if (props.controls) {
            sdk.hideControls();
            sdk.showControls(props.controls);
        }
        if (props.videoStyleObjectFit) {
            sdk.setVideoStyleObjectFit(props.videoStyleObjectFit);
        }
        if (props.videoStyleTransform) {
            sdk.setVideoStyleTransform(props.videoStyleTransform);
        }
        if (props.theme) {
            sdk.setTheme(props.theme);
        }
        sdk.addEventListener("ready", function () {
            sdk
                .getVolume()
                .then(function (d) { return props.onVolumeChange && props.onVolumeChange(d); });
            sdk
                .getDuration()
                .then(function (d) { return props.onDurationChange && props.onDurationChange(d); });
            sdk.getVideoSize().then(function (size) {
                var videoSizeRatio = size.height / size.width;
                if (props.onVideoSizeRatioChange) {
                    props.onVideoSizeRatioChange(videoSizeRatio);
                }
                setRatio(videoSizeRatio);
            });
        });
        setPlayerSdk(sdk);
    }, [iframeRef]);
    React.useImperativeHandle(ref, function () { return ({
        play: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.play(); },
        pause: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.pause(); },
        requestFullscreen: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.requestFullscreen(); },
        exitFullscreen: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.exitFullscreen(); },
        mute: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.mute(); },
        unmute: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.unmute(); },
        hideTitle: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hideTitle(); },
        showTitle: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showTitle(); },
        hideControls: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hideControls(); },
        showControls: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showControls(); },
        seek: function (time) {
            if (isNaN(time))
                throw new Error("Invalid time");
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.seek(time);
        },
        setCurrentTime: function (time) {
            if (isNaN(time))
                throw new Error("Invalid time");
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setCurrentTime(time);
        },
        setLoop: function (loop) { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setLoop(loop); },
        setPlaybackRate: function (rate) {
            if (isNaN(rate))
                throw new Error("Invalid rate");
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setPlaybackRate(rate);
        },
        setVolume: function (volume) {
            if (isNaN(volume))
                throw new Error("Invalid volume");
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setVolume(volume);
        },
        requestPictureInPicture: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.requestPictureInPicture(); },
        exitPictureInPicture: function () { return playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.exitPictureInPicture(); },
    }); });
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.loadConfig(__assign(__assign({}, props), props.video));
    }, [
        props.video.id,
        props.video.token,
        props.video.privateSession,
        props.video.live,
    ]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setChromeless(props.chromeless || false);
    }, [props.chromeless]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setLoop(props.loop || false);
    }, [props.loop]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setAutoplay(props.autoplay || false);
    }, [props.autoplay]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setMinimalQuality(props.minimalQuality || 0);
    }, [props.minimalQuality]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setMaximalQuality(props.maximalQuality || 0);
    }, [props.maximalQuality]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setVideoStyleTransform(props.videoStyleTransform || "");
    }, [props.videoStyleTransform]);
    React.useEffect(function () {
        playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setVideoStyleObjectFit(props.videoStyleObjectFit || "none");
    }, [props.videoStyleObjectFit]);
    React.useEffect(function () {
        if (props.muted !== undefined) {
            props.muted ? playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.mute() : playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.unmute();
        }
    }, [props.muted]);
    React.useEffect(function () {
        if (props.volume !== undefined) {
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setVolume(props.volume);
        }
    }, [props.volume]);
    React.useEffect(function () {
        if (props.hideTitle !== undefined) {
            props.hideTitle ? playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hideTitle() : playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showTitle();
        }
    }, [props.hideTitle]);
    React.useEffect(function () {
        if (props.hidePoster !== undefined) {
            props.hidePoster ? playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hidePoster() : playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showPoster();
        }
    }, [props.hidePoster]);
    React.useEffect(function () {
        if (props.showSubtitles !== undefined) {
            props.showSubtitles
                ? playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showSubtitles()
                : playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hideSubtitles();
        }
    }, [props.showSubtitles]);
    React.useEffect(function () {
        if (props.playbackRate !== undefined) {
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setPlaybackRate(props.playbackRate);
        }
    }, [props.playbackRate]);
    React.useEffect(function () {
        var allControls = [
            "play",
            "seekBackward",
            "seekForward",
            "playbackRate",
            "volume",
            "fullscreen",
            "subtitles",
            "chapters",
            "pictureInPicture",
            "progressBar",
            "chromecast",
            "download",
            "more",
        ];
        var oldControls = controls.length > 0 ? controls : allControls;
        var nextControls = props.controls != undefined ? props.controls : allControls;
        var added = nextControls.filter(function (c) { return oldControls.indexOf(c) === -1; });
        var removed = oldControls.filter(function (c) { return nextControls.indexOf(c) === -1; });
        if (added.length > 0)
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.showControls(added);
        if (removed.length > 0)
            playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.hideControls(removed);
        setControls(nextControls);
    }, [props.controls]);
    React.useEffect(function () {
        if (props.theme) {
            var themeHasChanged = function (oldTheme, newTheme) {
                return (Object.keys(oldTheme).length !== Object.keys(newTheme).length ||
                    Object.keys(oldTheme).filter(function (k) {
                        return oldTheme[k] !==
                            newTheme[k];
                    }).length > 0);
            };
            if (theme === null || themeHasChanged(theme, props.theme)) {
                playerSdk$1 === null || playerSdk$1 === void 0 ? void 0 : playerSdk$1.setTheme(props.theme);
                setTheme(props.theme);
            }
        }
    }, [props.theme]);
    var responsiveStyle = {};
    if (props.responsive) {
        if (ratio) {
            responsiveStyle.paddingTop = "".concat(ratio * 100, "%");
        }
        else {
            responsiveStyle.height = 0;
        }
        responsiveStyle.width = "100%";
    }
    return (React__namespace.createElement("div", { style: __assign(__assign(__assign({}, (props.style || {})), { position: "relative" }), responsiveStyle) },
        React__namespace.createElement("iframe", { allowFullScreen: true, allow: "autoplay; picture-in-picture", style: {
                width: "100%",
                height: "100%",
                border: "none",
                position: "absolute",
                top: 0,
                left: 0,
            }, ref: iframeRef }),
        props.children && (React__namespace.createElement("div", { style: {
                width: "100%",
                height: "100%",
                position: "absolute",
                top: 0,
            } }, props.children))));
});

module.exports = ApiVideoPlayer;
