// @api.video/nodejs-client@2.6.8/lib/model/AnalyticsAggregatedMetricsResponseContextTimeframe.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Returns the starting and ending date-times of the period you want analytics for.
 */
class AnalyticsAggregatedMetricsResponseContextTimeframe {
    static getAttributeTypeMap() {
        return AnalyticsAggregatedMetricsResponseContextTimeframe.attributeTypeMap;
    }
}
AnalyticsAggregatedMetricsResponseContextTimeframe.discriminator = undefined;
AnalyticsAggregatedMetricsResponseContextTimeframe.attributeTypeMap = [
    {
        name: 'from',
        baseName: 'from',
        type: 'Date',
        format: 'date-time',
    },
    {
        name: 'to',
        baseName: 'to',
        type: 'Date',
        format: 'date-time',
    },
];
exports.default = AnalyticsAggregatedMetricsResponseContextTimeframe;
