// @api.video/nodejs-client@2.6.8/lib/model/AnalyticsAggregatedMetricsResponse.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class AnalyticsAggregatedMetricsResponse {
    static getAttributeTypeMap() {
        return AnalyticsAggregatedMetricsResponse.attributeTypeMap;
    }
}
AnalyticsAggregatedMetricsResponse.discriminator = undefined;
AnalyticsAggregatedMetricsResponse.attributeTypeMap = [
    {
        name: 'context',
        baseName: 'context',
        type: 'AnalyticsAggregatedMetricsResponseContext',
        format: '',
    },
    {
        name: 'data',
        baseName: 'data',
        type: 'number',
        format: 'float',
    },
];
exports.default = AnalyticsAggregatedMetricsResponse;
