// @api.video/nodejs-client@2.6.8/lib/model/FilterBy.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class FilterBy {
    static getAttributeTypeMap() {
        return FilterBy.attributeTypeMap;
    }
}
FilterBy.discriminator = undefined;
FilterBy.attributeTypeMap = [
    {
        name: 'mediaId',
        baseName: 'mediaId',
        type: 'Array<string>',
        format: '',
    },
    {
        name: 'mediaType',
        baseName: 'mediaType',
        type: 'FilterByMediaTypeEnum',
        format: '',
    },
    {
        name: 'continent',
        baseName: 'continent',
        type: 'Array<FilterByContinentEnum>',
        format: '',
    },
    {
        name: 'country',
        baseName: 'country',
        type: 'Array<string>',
        format: '',
    },
    {
        name: 'deviceType',
        baseName: 'deviceType',
        type: 'Array<string>',
        format: '',
    },
    {
        name: 'operatingSystem',
        baseName: 'operatingSystem',
        type: 'Array<string>',
        format: '',
    },
    {
        name: 'browser',
        baseName: 'browser',
        type: 'Array<string>',
        format: '',
    },
    {
        name: 'tag',
        baseName: 'tag',
        type: 'string',
        format: '',
    },
    {
        name: 'referrer',
        baseName: 'referrer',
        type: 'Array<string>',
        format: 'uri',
    },
];
exports.default = FilterBy;
