// @api.video/nodejs-client@2.6.8/lib/model/SummaryCreationPayload.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class SummaryCreationPayload {
    static getAttributeTypeMap() {
        return SummaryCreationPayload.attributeTypeMap;
    }
}
SummaryCreationPayload.discriminator = undefined;
SummaryCreationPayload.attributeTypeMap = [
    {
        name: 'videoId',
        baseName: 'videoId',
        type: 'string',
        format: '',
    },
    {
        name: 'origin',
        baseName: 'origin',
        type: 'SummaryCreationPayloadOriginEnum',
        format: '',
    },
    {
        name: 'attributes',
        baseName: 'attributes',
        type: 'Array<SummaryCreationPayloadAttributesEnum>',
        format: '',
    },
];
exports.default = SummaryCreationPayload;
