// @api.video/nodejs-client@2.6.8/lib/model/Summary.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class Summary {
    static getAttributeTypeMap() {
        return Summary.attributeTypeMap;
    }
}
Summary.discriminator = undefined;
Summary.attributeTypeMap = [
    {
        name: 'summaryId',
        baseName: 'summaryId',
        type: 'string',
        format: '',
    },
    {
        name: 'createdAt',
        baseName: 'createdAt',
        type: 'Date',
        format: 'date-time',
    },
    {
        name: 'updatedAt',
        baseName: 'updatedAt',
        type: 'Date',
        format: 'date-time',
    },
    {
        name: 'videoId',
        baseName: 'videoId',
        type: 'string',
        format: '',
    },
    {
        name: 'origin',
        baseName: 'origin',
        type: 'SummaryOriginEnum',
        format: '',
    },
    {
        name: 'sourceStatus',
        baseName: 'sourceStatus',
        type: 'SummarySourceStatusEnum',
        format: '',
    },
];
exports.default = Summary;
