// @api.video/nodejs-client@2.6.8/lib/model/VideoClip.js
"use strict";
/**
 * @api.video/nodejs-client
 * api.video is an API that encodes on the go to facilitate immediate playback, enhancing viewer streaming experiences across multiple devices and platforms. You can stream live or on-demand online videos within minutes.
 *
 * The version of the OpenAPI document: 1
 *
 *
 * NOTE: This class is auto generated.
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Use this object to create a smaller clip from a video you upload.  - You can only create video clips in the same request where you create the video container. - You cannot update the starting or ending timestamps of a video clip after you created the video container. - When you upload a video file into a container where you defined a starting and ending timestamp, the API trims the video according to those timestamps to create a clip.
 */
class VideoClip {
    static getAttributeTypeMap() {
        return VideoClip.attributeTypeMap;
    }
}
VideoClip.discriminator = undefined;
VideoClip.attributeTypeMap = [
    {
        name: 'startTimecode',
        baseName: 'startTimecode',
        type: 'string',
        format: '',
    },
    {
        name: 'endTimecode',
        baseName: 'endTimecode',
        type: 'string',
        format: '',
    },
];
exports.default = VideoClip;
