// puppeteer-request-spy@1.4.0/build/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseModifier = exports.ResponseFaker = exports.RequestSpy = exports.RequestRedirector = exports.RequestModifier = exports.RequestInterceptor = void 0;
const RequestInterceptor_1 = require("./RequestInterceptor");
Object.defineProperty(exports, "RequestInterceptor", { enumerable: true, get: function () { return RequestInterceptor_1.RequestInterceptor; } });
const RequestModifier_1 = require("./RequestModifier");
Object.defineProperty(exports, "RequestModifier", { enumerable: true, get: function () { return RequestModifier_1.RequestModifier; } });
const RequestRedirector_1 = require("./RequestRedirector");
Object.defineProperty(exports, "RequestRedirector", { enumerable: true, get: function () { return RequestRedirector_1.RequestRedirector; } });
const RequestSpy_1 = require("./RequestSpy");
Object.defineProperty(exports, "RequestSpy", { enumerable: true, get: function () { return RequestSpy_1.RequestSpy; } });
const ResponseFaker_1 = require("./ResponseFaker");
Object.defineProperty(exports, "ResponseFaker", { enumerable: true, get: function () { return ResponseFaker_1.ResponseFaker; } });
const ResponseModifier_1 = require("./ResponseModifier");
Object.defineProperty(exports, "ResponseModifier", { enumerable: true, get: function () { return ResponseModifier_1.ResponseModifier; } });
